### PR TITLE
Story 2585 open source reporting

### PR DESCRIPTION
**Pull request by davidalk**
STORY-2585: DRR implementors can run reports based on open-source DSL

## Type of change

Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

### CODE REVIEW

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/GeneratorScope.java [hunk 1]
```diff
@@ -165,6 +177,23 @@ public GeneratedIdentifier getOrCreateIdentifier(RosettaNamed obj) {
...
+	 * @throws IllegalStateException if this scope already contains an identifier for `key`.
+	 */
+	public void createKeySynonym(Object key, Object keyWithIdentifier) {
+		if (isClosed) {
+			throw new IllegalStateException("Cannot create a new key synonym in a closed scope. (" + normalizeKey(key) + " -> " + normalizeKey(keyWithIdentifier) + ")\n" + this);
+		}
+		if (this.getIdentifier(key).isPresent()) {
+			throw new IllegalStateException("There is already a name defined for key `" + normalizeKey(key) + "`.\n" + this);
+		}
+
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/GeneratorScope.java:193]: Add check `keyWithIdentifier` has an identifier.

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/RosettaGenerator.xtend [hunk 2]
```diff
@@ -42,16 +49,19 @@ class RosettaGenerator implements IGenerator2 {
 
 	@Inject EnumGenerator enumGenerator
 	@Inject ModelMetaGenerator metaGenerator
-	@Inject DataRuleGenerator dataRuleGenerator
-	@Inject BlueprintGenerator blueprintGenerator
+	@Inject ConditionGenerator dataRuleGenerator
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/RosettaGenerator.xtend:52]: rename parameter as well

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/RosettaJavaPackages.java [hunk 3]
```diff
@@ -82,8 +82,16 @@ public DottedPath dataRule() {
 		public DottedPath existsValidation() {
 			return typeValidation().child("exists");
 		}
-
+		
 		public DottedPath blueprint() {
+ 			return child("blueprint");
+ 		}
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/RosettaJavaPackages.java:88]: Remove this one.

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/function/FunctionDependencyProvider.xtend [hunk 4]
```diff
@@ -59,21 +64,35 @@ class FunctionDependencyProvider {
...
 		}
 	}
-	
+
 	def Set<Function> functionDependencies(Iterable<? extends EObject> objects) {
-		distinctBy(objects.map[object | functionDependencies(object)].flatten, [f|f.name]).toSet;
+		distinctBy(objects.map[object|functionDependencies(object)].flatten, [f|f.name]).toSet;
+	}
+
+	def Set<RFunction> rFunctionDependencies(RosettaExpression expresseion) {
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/function/FunctionDependencyProvider.xtend:87]: fix typo

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/function/FunctionGenerator.xtend [hunk 5]
```diff
@@ -291,79 +339,73 @@ class FunctionGenerator {
...
+
+	private def JavaClass toDispatchClass(FunctionDispatch ele) {
+		return new JavaClass(DottedPath.splitOnDots(ele.model.name).child("functions"), ele.name + "." + ele.name + formatEnumName(ele.value.value.name))
 	}
-	
-	private def boolean assignAsKey(Operation op) {
+
+//	private def QualifiedName toEnumClassName(FunctionDispatch ele) {
+//		return QualifiedName.create(ele.name).append(formatEnumName(ele.value.value.name))
+//	}
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/function/FunctionGenerator.xtend:366]: Remove this

File: rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/types/JavaTypeTranslator.java [hunk 6]
```diff
@@ -73,26 +91,60 @@ private DottedPath getModelPackage(RosettaNamed object) {
...
+	public JavaClass toFunctionJavaClass(RFunction func) {
+		switch (func.getOrigin()) {
+		case FUNCTION:
+			return generatedJavaClassService.toJavaFunction(func.getModelSymbolId());
+		case REPORT:
+			return generatedJavaClassService.toJavaReportFunction(func.getModelSymbolId());
+		case RULE:
+			return generatedJavaClassService.toJavaRule(func.getModelSymbolId());
+		default:
+			throw new IllegalStateException();
```
Comment (SimonCockx) [inline rosetta-lang/src/main/java/com/regnosys/rosetta/generator/java/types/JavaTypeTranslator.java:110]: Add origin info in the exception.

File: rosetta-testing/src/test/java/com/regnosys/rosetta/generator/java/object/ModelMetaGeneratorTest.xtend [hunk 7]
```diff
@@ -142,9 +142,9 @@ class ModelMetaGeneratorTest {
 						).stream().filter(res -> !res.get()).map(res -> res.getError()).collect(joining("; "));
 					
 					if (!isNullOrEmpty(error)) {
-						return failure("Foo", ValidationResult.ValidationType.CARDINALITY, o.getClass().getSimpleName(), path, "", error);
+						return failure("Foo", ValidationType.CARDINALITY, o.getClass().getSimpleName(), path, "", error);
```
Comment (davidalk) [inline rosetta-testing/src/test/java/com/regnosys/rosetta/generator/java/object/ModelMetaGeneratorTest.xtend:145]: Should be `o.getType()`

File: rosetta-testing/src/test/java/com/regnosys/rosetta/generator/java/reports/TabulatorTest.xtend [hunk 8]
```diff
@@ -0,0 +1,831 @@
+package com.regnosys.rosetta.generator.java.reports
+
+import com.google.inject.Inject
```
Comment (davidalk) [inline rosetta-testing/src/test/java/com/regnosys/rosetta/generator/java/reports/TabulatorTest.xtend:3]: Should be `javax`

**Review by SimonCockx — APPROVED** [review]

**Pull request merged**